### PR TITLE
Add TypeScript guide

### PR DESCRIPTION
**Pull request by dhruvdutt**
image

### CODE REVIEW

**Comment by dhruvdutt** [comment]
I think we should keep [#223](https://github.com/appbaseio/reactivesearch/issues/223) open (or file a new issue) as an opportunity for outsiders to contribute.
There are still lots of improvements that can be done in the TypeScript definitions.

File: content/docs/advanced/typescript.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+permalink: advanced/typescript.html
+prev: advanced/ssr.html
+prevTitle: "Server Side Rendering"
+redirect_from:
+    - "typescript"
+    - "advanced/typescript"
+---
+
+
+Add static typing to JavaScript to improve developer productivity and code quality with <a href="https://www.typescriptlang.org/" target="_blank">TypeScript</a>.
```
Comment (divyanshu013) [inline content/docs/advanced/typescript.md:15]: This line can be removed I think.

**Comment by dhruvdutt** [comment]
@divyanshu013 Improvements in terms of methods and callback signatures. For instance, shape of `onAllData` along with interfaces for args like `streamResults`. Currently, we're checking for only types of props (boolean, string, func, array) and allow any types of functions.

One more example would be that things like `onKeyPressDown` are `KeyboardEventHandler<T>`; `onChange` is `FormEventHandler<T>`. All these are types of [SyntheticEvent](https://reactjs.org/docs/events.html#mouse-events).

Implementing this is a lot of work (since we don't use TS internally in source) so I and @metagrover agreed to keep initial support limited to types only.

File: content/docs/advanced/typescript.md (hunk 1 above)
Comment (dhruvdutt) [reply content/docs/advanced/typescript.md:15]: One line intro for those who stumble into TypeScript guide. I think if we don't have much content, it's okay to keep it.

Comment (divyanshu013) [reply content/docs/advanced/typescript.md:15]: I'm a bit confused about
> improve developer productivity and code quality

**Review by siddharthlatest — APPROVED** [review]

**Pull request merged**